/*
 * Copyright (C) 2017-2018 GIG Technology NV and Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Use `go generate ./daemon/api/grpc`
// to generate the corresponding `.pb.go` file.
syntax = "proto3";

package schema;

// FileService is the service to be used to write, read, delete, check and repair files.
// The fileService follows the principle of everything is a file.
// All files are written as raw binary data, but also have
// metadata bound to them, which identify where and how they are stored.
// This metadata is returned, when creating or altering the metadata,
// however if not needed for anything it can be ignored, as it is stored by the daemon (server),
// already, using the underlying metastor client.
service FileService {
	rpc Write(WriteRequest) returns (WriteResponse) {}
	rpc WriteFile(WriteFileRequest) returns (WriteFileResponse) {}
	rpc WriteStream(stream WriteStreamRequest) returns (WriteStreamResponse) {}

	rpc Read(ReadRequest) returns (ReadResponse) {}
	rpc ReadFile(ReadFileRequest) returns (ReadFileResponse) {}
	rpc ReadStream(ReadStreamRequest) returns (stream ReadStreamResponse) {}

	rpc Delete(DeleteRequest) returns (DeleteResponse) {}

	rpc Check(CheckRequest) returns (CheckResponse) {}
	rpc Repair(RepairRequest) returns (RepairResponse) {}
}

message Metadata {
    // key defines the key of the data,
    // and is chosen by the owner of this data.
    bytes key = 1;

    // totalSize in bytes represents the total size of all chunks,
    // that make up the stored data, combined.
    int64 totalSize = 2;

    // creationEpoch defines the time this data was initially created,
    // in the Unix epoch format, in nano seconds.
    int64 creationEpoch = 3;
    // lastWriteEpoch defines the time this data
    // was last modified (e.g. repaired),
    // in the Unix epoch format, in nano seconds.
    int64 lastWriteEpoch = 4;

    // chunks is the metadata list of all chunks
    // that make up the data, when combined.
    repeated Chunk chunks = 5;
}
message Chunk {
    // chunkSize of the chunk in bytes
    int64 chunkSize = 1;

    // objects defines the metadata of the objects
    // that make up this chunk.
    repeated Object objects = 2;

    // hash contains the checksum/signature of the chunk (data),
    // meaning the data of all objects (of this chunk) combined.
    bytes hash = 3;
}
message Object {
    // key of the Object
    bytes key = 1;

    // shardID defines the ID of the shard the object is stored on
    string shardID = 2;
}

enum CheckStatus {
	CheckStatusInvalid = 0;
	CheckStatusValid = 1;
	CheckStatusOptimal = 2;
}

enum FileMode {
	// if the file already exists, truncate it to 0 bytes prior to writing
	FileModeTruncate = 0;
	// append to the file should it already exists, otherwise create it
	FileModeAppend = 1;
	// create a non-existing file and write to it
	FileModeExclusive = 2;
}

message WriteRequest {
	bytes key = 1;
	bytes data = 2;
}
message WriteResponse {
	Metadata metadata = 1;
}

message WriteFileRequest {
	bytes key = 1;
	string filePath = 2;
}
message WriteFileResponse {
	Metadata metadata = 1;
}

// key is send as part of header
message WriteStreamRequest {
	oneof input {
		Metadata metadata = 1; // first message should be the metadata
		Data data = 2; // all other messages (at least one) should be data itself
	}

	message Metadata {
		bytes key = 1;
	}
	message Data {
		bytes dataChunk = 2;
	}
}
message WriteStreamResponse {
	Metadata metadata = 1;
}

message ReadRequest {
	oneof input {
		bytes key = 1;
		Metadata metadata = 2;
	}
}
message ReadResponse {
	bytes data = 1;
}

message ReadFileRequest {
	oneof input {
		bytes key = 1;
		Metadata metadata = 2;
	}

	// destination file and its configuration
	string filePath = 3;
	FileMode fileMode = 4;
	bool synchronousIO = 5; // open file for synchronous I/O

}
message ReadFileResponse {
}

message ReadStreamRequest {
	oneof input {
		bytes key = 1;
		Metadata metadata = 2;
	}
	int64 chunkSize = 3;
}
message ReadStreamResponse {
	bytes dataChunk = 1;
}

message DeleteRequest {
	oneof input {
		bytes key = 1;
		Metadata metadata = 2;
	}
}
message DeleteResponse {
}

message CheckRequest {
	oneof input {
		bytes key = 1;
		Metadata metadata = 2;
	}
	bool fast = 3;
}
message CheckResponse {
	CheckStatus status = 1;
}

message RepairRequest {
	bytes key = 1;
}
message RepairResponse {
	Metadata metadata = 1;
}

// MetadataService is used to set, get and delete metadata directly
// using the underlying metadata storage client.
service MetadataService {
	rpc SetMetadata(SetMetadataRequest) returns (SetMetadataResponse) {}
	rpc GetMetadata(GetMetadataRequest) returns (GetMetadataResponse) {}
	rpc DeleteMetadata(DeleteMetadataRequest) returns (DeleteMetadataResponse) {}
	rpc ListKeys(ListMetadataKeysRequest) returns (stream ListMetadataKeysResponse) {}

	// TODO:
	// figure out a way to support UpdateMetadata,
	// as this would be a feature required,
	// should you want to (partially) update existing
	// metadata
}

message SetMetadataRequest {
	Metadata metadata = 1;
}
message SetMetadataResponse {
}

message GetMetadataRequest {
	bytes key = 1;
}
message GetMetadataResponse {
	Metadata metadata = 1;
}

message DeleteMetadataRequest {
	bytes key = 1;
}
message DeleteMetadataResponse {
}

message ListMetadataKeysRequest {
}

message ListMetadataKeysResponse {
	bytes key = 1;
}

// DataService is used write, read, delete, check and repair (processed) data.
// as data can be written to multiple servers and/or be split over multiple chunks,
// some metadata is returned which the user is expected to store somewhere,
// as that metadata is required to read and manage the data, later on.
//
// Should you not want to have to deal with this metadata yourself,
// you should use the FileService instead, which is one level higher,
// and takes care of managing and storing the metadata for you,
// as well as generating extra metadata, such as the creation epoch time.
service DataService {
	rpc Write(DataWriteRequest) returns (DataWriteResponse) {}
	rpc WriteFile(DataWriteFileRequest) returns (DataWriteFileResponse) {}
	rpc WriteStream(stream DataWriteStreamRequest) returns (DataWriteStreamResponse) {}

	rpc Read(DataReadRequest) returns (DataReadResponse) {}
	rpc ReadFile(DataReadFileRequest) returns (DataReadFileResponse) {}
	rpc ReadStream(DataReadStreamRequest) returns (stream DataReadStreamResponse) {}

	rpc Delete(DataDeleteRequest) returns (DataDeleteResponse) {}

	rpc Check(DataCheckRequest) returns (DataCheckResponse) {}
	rpc Repair(DataRepairRequest) returns (DataRepairResponse) {}
}

message DataWriteRequest {
	bytes data = 1;
}
message DataWriteResponse {
	repeated Chunk chunks = 1;
}

message DataWriteFileRequest {
	string filePath = 1;
}
message DataWriteFileResponse {
	repeated Chunk chunks = 1;
}

message DataWriteStreamRequest {
	bytes dataChunk = 1;
}
message DataWriteStreamResponse {
	repeated Chunk chunks = 1;
}

message DataReadRequest {
	repeated Chunk chunks = 1;
}
message DataReadResponse {
	bytes data = 1;
}

message DataReadFileRequest {
	repeated Chunk chunks = 1;
	// destination file and its configuration
	string filePath = 2;
	FileMode fileMode = 3;
	bool synchronousIO = 4; // open file for synchronous I/O
}
message DataReadFileResponse {
}

message DataReadStreamRequest {
	repeated Chunk chunks = 1;
	int64 chunkSize = 2;
}
message DataReadStreamResponse {
	bytes dataChunk = 1;
}

message DataDeleteRequest {
	repeated Chunk chunks = 1;
}
message DataDeleteResponse {
}

message DataCheckRequest {
	repeated Chunk chunks = 1;
	bool fast = 2;
}
message DataCheckResponse {
	CheckStatus status = 1;
}

message DataRepairRequest {
	repeated Chunk chunks = 1;
}
message DataRepairResponse {
	repeated Chunk chunks = 1;
}
